config.yaml

```diff
@@ -1,14 +1,42 @@
-# This file defines charm config options, and populates the Configure tab on Charmhub.
-# If your charm does not require configuration options, delete this file entirely.
-#
-# See https://juju.is/docs/config for guidance.
-
 options:
-  role:
-    description: Role for this microk8s deployment
-    default: ""
+  addons:
+    description: |
+      Space-separated list of addons that should be enabled and managed by the charm. Removing
+      an addon from this list will disable it in the cluster.
+
+      See https://microk8s.io/docs/addons for a comprehensive list of addons available in MicroK8s.
+
+      Use "addon:arg1:arg2" to specify arguments when enabling a specific addon.
+
+      Examples:
+      - "dns rbac hostpath-storage ingress"
+      - "dns rbac hostpath-storage ingress gpu:--version=v1.10.1"
+
+    default: "dns rbac hostpath-storage ingress"
     type: string
   channel:
-    description: MicroK8s channel to install
+    description: |
+      Specify MicroK8s version to install. This is the channel that will be passed to the "snap
+      install microk8s" command. See also https://microk8s.io/docs/setting-snap-channel
+
+      Examples:
+      - ""                  # install latest version
+      - "1.27/stable"       # install MicroK8s 1.27 from the stable branch
+      - "1.27/edge"         # install MicroK8s 1.27 from the edge branch (getting all latest fixes)
+      - "1.27-strict"       # install MicroK8s 1.27 with strict confinement
+    default: ""
+    type: string
+  role:
```

Comment:
This should probably be derived from relations if at all possible

Reply:
Yes, this was in fact heavily considered initially. However, deriving this from whether a relation exists or not seems to be way too late and/or not clear. Imagine the following scenario:

```bash
# intent: deploy a 3 node control plane
juju deploy microk8s microk8s -n 3

# intent: deploy 3 nodes to join as workers
juju deploy microk8s microk8s-worker -n 3

# intent: join 3 workers
juju integrate microk8s:microk8s-provides microk8s-worker:microk8s
```

By the time we integrate those charms, the `microk8s-worker` application might have already started clustering itself using the `peer` relation as a 3-node control plane, enabling addons, etc. there's quite a lot of extra steps to take in this scenario to undo things and it feels unnecessary.

Further, it is not entirely clear what the steps to follow would be in the scenario where more units are added, or some units are removed. Having a set `role` in this case makes things much clearer and the intent explicit, e.g.:

```bash
# intent: deploy a 3-node control plane
juju deploy microk8s microk8s -n 3

# intent: deploy 3 nodes to join as workers
juju deploy microk8s microk8s-worker -n 3 --config role=worker

# intent: join the cluster as workers
integrate microk8s:microk8s-provides microk8s-worker:microk8s
```

More than glad to explore this further in a future iteration, but for now it allows us to focus on more important tasks.

Reply:
Okay, I actually now remember talking about this in Prague. On reflection, this doesn't seem so bad, and will indeed allow you to move a little quicker.